File: app/src/main/java/com/sakhand/downloadmanager/util/Format.kt
package com.sakhand.downloadmanager.util

/**
 * ابزارهای قالب‌بندی اعداد و حجم به سبک فارسی
 */

private val FA_DIGITS = mapOf(
    '0' to '۰', '1' to '۱', '2' to '۲', '3' to '۳', '4' to '۴',
    '5' to '۵', '6' to '۶', '7' to '۷', '8' to '۸', '9' to '۹',
    '.' to '٫', ',' to '٬'
)

fun toFa(value: Long): String = value.toString().map { FA_DIGITS[it] ?: it }.joinToString("")

fun toFa(value: Int): String = toFa(value.toLong())

fun String.toFaDigits(): String = map { FA_DIGITS[it] ?: it }.joinToString("")

fun formatBytes(bytes: Long): String {
    if (bytes < 0) return "نامشخص"
    if (bytes == 0L) return "۰ بایت"
    val b = bytes.toDouble()
    return when {
        b >= 1024.0 * 1024 * 1024 -> String.format(java.util.Locale.US, "%.2f گیگابایت", b / (1024.0 * 1024 * 1024))
        b >= 1024.0 * 1024 -> String.format(java.util.Locale.US, "%.1f مگابایت", b / (1024.0 * 1024))
        b >= 1024.0 -> String.format(java.util.Locale.US, "%.1f کیلوبایت", b / 1024.0)
        else -> "$bytes بایت"
    }.toFaDigits()
}

fun formatSpeed(bps: Long): String {
    if (bps <= 0) return "—"
    return formatBytes(bps) + " بر ثانیه"
}

fun etaText(remainingBytes: Long, bps: Long): String {
    if (bps <= 0 || remainingBytes <= 0) return "—"
    val sec = remainingBytes / bps
    return when {
        sec < 60 -> "حدود ${toFa(sec)} ثانیه"
        sec < 3600 -> "حدود ${toFa(sec / 60)} دقیقه"
        else -> "حدود ${toFa(sec / 3600)} ساعت"
    }
}

File: windows/SakhandDm/Core/Format.cs
using System.Collections.Generic;

namespace SakhandDm.Core;

/// <summary>قالب‌بندی اعداد و حجم به سبک فارسی</summary>
public static class Fmt
{
    private static readonly Dictionary<char, char> FaDigits = new()
    {
        ['0'] = '۰', ['1'] = '۱', ['2'] = '۲', ['3'] = '۳', ['4'] = '۴',
        ['5'] = '۵', ['6'] = '۶', ['7'] = '۷', ['8'] = '۸', ['9'] = '۹',
        ['.'] = '٫', [','] = '٬'
    };

    public static string ToFa(long value)
    {
        var s = value.ToString();
        var sb = new System.Text.StringBuilder(s.Length);
        foreach (var c in s) sb.Append(FaDigits.TryGetValue(c, out var f) ? f : c);
        return sb.ToString();
    }

    public static string ToFaDigits(this string s)
    {
        var sb = new System.Text.StringBuilder(s.Length);
        foreach (var c in s) sb.Append(FaDigits.TryGetValue(c, out var f) ? f : c);
        return sb.ToString();
    }

    public static string FormatBytes(long bytes)
    {
        if (bytes < 0) return "نامشخص";
        if (bytes == 0) return "۰ بایت";
        double b = bytes;
        string s = b >= 1024d * 1024 * 1024
            ? string.Format(System.Globalization.CultureInfo.InvariantCulture, "{0:F2} گیگابایت", b / (1024d * 1024 * 1024))
            : b >= 1024d * 1024
                ? string.Format(System.Globalization.CultureInfo.InvariantCulture, "{0:F1} مگابایت", b / (1024d * 1024))
                : b >= 1024
                    ? string.Format(System.Globalization.CultureInfo.InvariantCulture, "{0:F1} کیلوبایت", b / 1024)
                    : $"{bytes} بایت";
        return s.ToFaDigits();
    }

    public static string FormatSpeed(long bps) => bps <= 0 ? "—" : FormatBytes(bps) + " بر ثانیه";

    public static string EtaText(long remainingBytes, long bps)
    {
        if (bps <= 0 || remainingBytes <= 0) return "—";
        long sec = remainingBytes / bps;
        if (sec < 60) return $"حدود {ToFa(sec)} ثانیه";
        if (sec < 3600) return $"حدود {ToFa(sec / 60)} دقیقه";
        return $"حدود {ToFa(sec / 3600)} ساعت";
    }
}
